test_db_nfs_mount - There are persistent failures of this test case due to ResourceWrongStatusException , fix of the test ix needed

test_db_nfs_mount - There are persistent failures of this test case due to ResourceWrongStatusException , fix of the test ix needed. 
https://reportportal-ocs4.apps.ocp-c1.prod.psi.redhat.com/ui/#ocs/launches/795/28022/1376497/1376541/log
https://reportportal-ocs4.apps.ocp-c1.prod.psi.redhat.com/ui/#ocs/launches/795/27570/1359553/1359607/log
https://reportportal-ocs4.apps.ocp-c1.prod.psi.redhat.com/ui/#ocs/launches/795/27647/1363270/1363314/log

etc 

The relevant backtrace:

resource = <ocs_ci.ocs.resources.ocs.OCS object at 0x7f08d7c167f0>
state = 'Running', timeout = 60


def wait_for_resource_state(resource, state, timeout=60):
    """
    Wait for a resource to get to a given status

    Args:
        resource (OCS obj): The resource object
        state (str): The status to wait for
        timeout (int): Time in seconds to wait

    Raises:
        ResourceWrongStatusException: In case the resource hasn't
            reached the desired state

    """
    if check_cluster_is_compact():
        timeout = 180
    if (
        resource.name == constants.DEFAULT_STORAGECLASS_CEPHFS
        or resource.name == constants.DEFAULT_STORAGECLASS_RBD
    ):
        logger.info("Attempt to default default Secret or StorageClass")
        return
    try:

      resource.ocp.wait_for_resource(
            condition=state, resource_name=resource.name, timeout=timeout
        )

/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/helpers/helpers.py:135: 



self = <ocs_ci.ocs.ocp.OCP object at 0x7f08a1ed99d0>, condition = 'Running'
resource_name = 'demo-pod', column = 'STATUS', selector = None
resource_count = 0, timeout = 60, sleep = 3, dont_allow_other_resources = False
error_condition = None


def wait_for_resource(
    self,
    condition,
    resource_name="",
    column="STATUS",
    selector=None,
    resource_count=0,
    timeout=60,
    sleep=3,
    dont_allow_other_resources=False,
    error_condition=None,
):
    """
    Wait for a resource to reach to a desired condition

    Args:
        condition (str): The desired state the resource that is sampled
            from 'oc get <kind> <resource_name>' command
        resource_name (str): The name of the resource to wait
            for (e.g.my-pv1)
        column (str): The name of the column to compare with
        selector (str): The resource selector to search with.
            Example: 'app=rook-ceph-mds'
        resource_count (int): How many resources expected to be
        timeout (int): Time in seconds to wait
        sleep (int): Sampling time in seconds
        dont_allow_other_resources (bool): If True it will not allow other
            resources in different state. For example you are waiting for 2
            resources and there are currently 3 (2 in running state,
            1 in ContainerCreating) the function will continue to next
            iteration to wait for only 2 resources in running state and no
            other exists.
        error_condition (str): State of the resource that is sampled
            from 'oc get <kind> <resource_name>' command, which makes this
            method to fail immediately without waiting for a timeout. This
            is optional and makes sense only when there is a well defined
            unrecoverable state of the resource(s) which is not expected to
            be part of a workflow under test, and at the same time, the
            timeout itself is large.

    Returns:
        bool: True in case all resources reached desired condition,
            False otherwise

    """
    if condition == error_condition:
        # when this fails, this method is used in a wrong way
        raise ValueError(
            f"Condition '{condition}' we are waiting for must be different"
            f" from error condition '{error_condition}'"
            " which describes unexpected error state."
        )
    log.info(
        (
            f"Waiting for a resource(s) of kind {self._kind}"
            f" identified by name '{resource_name}'"
            f" using selector {selector}"
            f" at column name {column}"
            f" to reach desired condition {condition}"
        )
    )
    resource_name = resource_name if resource_name else self.resource_name
    selector = selector if selector else self.selector

    # actual status of the resource we are waiting for, setting it to None
    # now prevents UnboundLocalError raised when waiting timeouts
    actual_status = None

    try:
        for sample in TimeoutSampler(
            timeout, sleep, self.get, resource_name, True, selector
        ):
            # Only 1 resource expected to be returned
            if resource_name:
                retry = int(timeout / sleep if sleep else timeout / 1)
                status = self.get_resource(
                    resource_name,
                    column,
                    retry=retry,
                    wait=sleep,
                )
                if status == condition:
                    log.info(
                        f"status of {resource_name} at {column}"
                        " reached condition!"
                    )
                    return True
                log.info(
                    (
                        f"status of {resource_name} at column {column} was {status},"
                        f" but we were waiting for {condition}"
                    )
                )
                actual_status = status
                if error_condition is not None and status == error_condition:
                    raise ResourceWrongStatusException(
                        resource_name,
                        column=column,
                        expected=condition,
                        got=status,
                    )
            # More than 1 resources returned
            elif sample.get("kind") == "List":
                in_condition = []
                in_condition_len = 0
                actual_status = []
                sample = sample["items"]
                sample_len = len(sample)
                for item in sample:
                    try:
                        item_name = item.get("metadata").get("name")
                        status = self.get_resource(item_name, column)
                        actual_status.append(status)
                        if status == condition:
                            in_condition.append(item)
                            in_condition_len = len(in_condition)
                        if (
                            error_condition is not None
                            and status == error_condition
                        ):
                            raise ResourceWrongStatusException(
                                item_name,
                                column=column,
                                expected=condition,
                                got=status,
                            )
                    except CommandFailed as ex:
                        log.info(
                            f"Failed to get status of resource: {item_name} at column {column}, "
                            f"Error: {ex}"
                        )
                    if resource_count:
                        if in_condition_len == resource_count:
                            log.info(
                                f"{in_condition_len} resources already "
                                f"reached condition!"
                            )
                            if (
                                dont_allow_other_resources
                                and sample_len != in_condition_len
                            ):
                                log.info(
                                    f"There are {sample_len} resources in "
                                    f"total. Continue to waiting as "
                                    f"you don't allow other resources!"
                                )
                                continue
                            return True
                    elif len(sample) == len(in_condition):
                        return True
                # preparing logging message with expected number of
                # resource items we are waiting for
                if resource_count > 0:
                    exp_num_str = f"all {resource_count}"
                else:
                    exp_num_str = "all"
                log.info(
                    (
                        f"status of {resource_name} at column {column} - item(s) were {actual_status},"
                        f" but we were waiting"
                        f" for {exp_num_str} of them to be {condition}"
                    )
                )
    except TimeoutExpiredError as ex:
        log.error(f"timeout expired: {ex}")
        # run `oc describe` on the resources we were waiting for to provide
        # evidence so that we can understand what was wrong
        output = self.describe(resource_name, selector=selector)
        log.warning(
            "Description of the resource(s) we were waiting for:\n%s", output
        )
        log.error(
            (
                f"Wait for {self._kind} resource {resource_name} at column {column}"
                f" to reach desired condition {condition} failed,"
                f" last actual status was {actual_status}"
            )
        )

      raise (ex)

/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/ocs/ocp.py:873: 



self = <ocs_ci.ocs.ocp.OCP object at 0x7f08a1ed99d0>, condition = 'Running'
resource_name = 'demo-pod', column = 'STATUS', selector = None
resource_count = 0, timeout = 60, sleep = 3, dont_allow_other_resources = False
error_condition = None


def wait_for_resource(
    self,
    condition,
    resource_name="",
    column="STATUS",
    selector=None,
    resource_count=0,
    timeout=60,
    sleep=3,
    dont_allow_other_resources=False,
    error_condition=None,
):
    """
    Wait for a resource to reach to a desired condition

    Args:
        condition (str): The desired state the resource that is sampled
            from 'oc get <kind> <resource_name>' command
        resource_name (str): The name of the resource to wait
            for (e.g.my-pv1)
        column (str): The name of the column to compare with
        selector (str): The resource selector to search with.
            Example: 'app=rook-ceph-mds'
        resource_count (int): How many resources expected to be
        timeout (int): Time in seconds to wait
        sleep (int): Sampling time in seconds
        dont_allow_other_resources (bool): If True it will not allow other
            resources in different state. For example you are waiting for 2
            resources and there are currently 3 (2 in running state,
            1 in ContainerCreating) the function will continue to next
            iteration to wait for only 2 resources in running state and no
            other exists.
        error_condition (str): State of the resource that is sampled
            from 'oc get <kind> <resource_name>' command, which makes this
            method to fail immediately without waiting for a timeout. This
            is optional and makes sense only when there is a well defined
            unrecoverable state of the resource(s) which is not expected to
            be part of a workflow under test, and at the same time, the
            timeout itself is large.

    Returns:
        bool: True in case all resources reached desired condition,
            False otherwise

    """
    if condition == error_condition:
        # when this fails, this method is used in a wrong way
        raise ValueError(
            f"Condition '{condition}' we are waiting for must be different"
            f" from error condition '{error_condition}'"
            " which describes unexpected error state."
        )
    log.info(
        (
            f"Waiting for a resource(s) of kind {self._kind}"
            f" identified by name '{resource_name}'"
            f" using selector {selector}"
            f" at column name {column}"
            f" to reach desired condition {condition}"
        )
    )
    resource_name = resource_name if resource_name else self.resource_name
    selector = selector if selector else self.selector

    # actual status of the resource we are waiting for, setting it to None
    # now prevents UnboundLocalError raised when waiting timeouts
    actual_status = None

    try:

      for sample in TimeoutSampler(
            timeout, sleep, self.get, resource_name, True, selector
        ):

/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/ocs/ocp.py:764: 



self = <ocs_ci.utility.utils.TimeoutSampler object at 0x7f08d7dbec10>


def __iter__(self):
    if self.start_time is None:
        self.start_time = time.time()
    while True:
        self.last_sample_time = time.time()
        if self.timeout <= (self.last_sample_time - self.start_time):

          raise self.timeout_exc_cls(*self.timeout_exc_args)
E               ocs_ci.ocs.exceptions.TimeoutExpiredError: Timed out after 60s running get("demo-pod", True, None)



/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/utility/utils.py:1481: TimeoutExpiredError


During handling of the above exception, another exception occurred:


self = <mcg.test_db_nfs_mount.TestNoobaaDbNFSMount object at 0x7f08a2c84ca0>
request = <SubRequest 'mount_ngix_pod' for >


@pytest.fixture()
def mount_ngix_pod(self, request):
    # try to mount the reesi004 nfs mount to nginx pod
    nginx_pod_data = templating.load_yaml(constants.NGINX_POD_YAML)
    nginx_pod_data["metadata"]["namespace"] = config.ENV_DATA["cluster_namespace"]
    nginx_pod_data["spec"]["containers"][0]["volumeMounts"][0]["name"] = "nfs-vol"
    nginx_pod_data["spec"]["containers"][0]["volumeMounts"][0][
        "mountPath"
    ] = "/var/nfs"
    nginx_pod_data["spec"]["volumes"][0]["name"] = "nfs-vol"
    nginx_pod_data["spec"]["volumes"][0]["nfs"] = dict()
    nginx_pod_data["spec"]["volumes"][0]["nfs"]["server"] = config.ENV_DATA.get(
        "nb_nfs_server"
    )
    nginx_pod_data["spec"]["volumes"][0]["nfs"]["path"] = config.ENV_DATA.get(
        "nb_nfs_mount"
    )
    nginx_pod_data["spec"]["volumes"][0].pop("persistentVolumeClaim")
    nginx_pod = helpers.create_resource(**nginx_pod_data)

  helpers.wait_for_resource_state(
        nginx_pod,
        constants.STATUS_RUNNING,
    )

/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/tests/functional/object/mcg/test_db_nfs_mount.py:49: 



resource = <ocs_ci.ocs.resources.ocs.OCS object at 0x7f08d7c167f0>
state = 'Running', timeout = 60


def wait_for_resource_state(resource, state, timeout=60):
    """
    Wait for a resource to get to a given status

    Args:
        resource (OCS obj): The resource object
        state (str): The status to wait for
        timeout (int): Time in seconds to wait

    Raises:
        ResourceWrongStatusException: In case the resource hasn't
            reached the desired state

    """
    if check_cluster_is_compact():
        timeout = 180
    if (
        resource.name == constants.DEFAULT_STORAGECLASS_CEPHFS
        or resource.name == constants.DEFAULT_STORAGECLASS_RBD
    ):
        logger.info("Attempt to default default Secret or StorageClass")
        return
    try:
        resource.ocp.wait_for_resource(
            condition=state, resource_name=resource.name, timeout=timeout
        )
    except TimeoutExpiredError:
        logger.error(f"{resource.kind} {resource.name} failed to reach {state}")
        resource.reload()

      raise ResourceWrongStatusException(resource.name, resource.describe())
E           ocs_ci.ocs.exceptions.ResourceWrongStatusException: Resource demo-pod describe output: Name:             demo-pod
E           Namespace:        openshift-storage
E           Priority:         0
E           Service Account:  default
E           Node:             compute-1/10.1.112.151
E           Start Time:       Thu, 16 Jan 2025 07:48:58 -0500
E           Labels:           
E           Annotations:      k8s.ovn.org/pod-networks:
E                               {"default":{"ip_addresses":["10.128.2.177/23"],"mac_address":"0a:58:0a:80:02:b1","gateway_ips":["10.128.2.1"],"routes":[{"dest":"10.128.0....
E                             openshift.io/scc: noobaa-core
E           Status:           Pending
E           IP:
E           IPs:              
E           Containers:
E             web-server:
E               Container ID:
E               Image:          ocs-qe-proxy-dc-eco.usersys.redhat.com:5000/ocsci/nginx:fio
E               Image ID:
E               Port:           
E               Host Port:      
E               State:          Waiting
E                 Reason:       ContainerCreating
E               Ready:          False
E               Restart Count:  0
E               Environment:    
E               Mounts:
E                 /var/nfs from nfs-vol (rw)
E                 /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-86cbg (ro)
E           Conditions:
E             Type                        Status
E             PodReadyToStartContainers   False
E             Initialized                 True
E             Ready                       False
E             ContainersReady             False
E             PodScheduled                True
E           Volumes:
E             nfs-vol:
E               Type:      NFS (an NFS mount that lasts the lifetime of a pod)
E               Server:    odf-ci-test-nfs-server.usersys.redhat.com
E               Path:      /noobaa_test/
E               ReadOnly:  false
E             kube-api-access-86cbg:
E               Type:                    Projected (a volume that contains injected data from multiple sources)
E               TokenExpirationSeconds:  3607
E               ConfigMapName:           kube-root-ca.crt
E               ConfigMapOptional:       
E               DownwardAPI:             true
E               ConfigMapName:           openshift-service-ca.crt
E               ConfigMapOptional:       
E           QoS Class:                   BestEffort
E           Node-Selectors:              
E           Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
E                                        node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
E           Events:
E             Type    Reason     Age   From               Message
E             ----    ------     ----  ----               -------
E             Normal  Scheduled  64s   default-scheduler  Successfully assigned openshift-storage/demo-pod to compute-1



/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/helpers/helpers.py:141: ResourceWrongStatusException

--- a/tests/functional/object/mcg/test_db_nfs_mount.py
+++ b/tests/functional/object/mcg/test_db_nfs_mount.py
@@ -28,13 +28,18 @@ logger = logging.getLogger(__name__)
 @skipif_lean_deployment
 class TestNoobaaDbNFSMount:
     @pytest.fixture()
-    def mount_ngix_pod(self, request):
+    def nfs_config(self):
         nfs_server = config.ENV_DATA.get("nb_nfs_server")
         nfs_mount = config.ENV_DATA.get("nb_nfs_mount")
         if not nfs_server or not nfs_mount:
             pytest.skip(
                 f"NFS config not set: nb_nfs_server={nfs_server}, nb_nfs_mount={nfs_mount}"
             )
+        return nfs_server, nfs_mount
+
+    @pytest.fixture()
+    def mount_ngix_pod(self, request, nfs_config):
+        nfs_server, nfs_mount = nfs_config
 
         # try to mount the reesi004 nfs mount to nginx pod
         nginx_pod_data = templating.load_yaml(constants.NGINX_POD_YAML)
@@ -64,13 +69,8 @@ class TestNoobaaDbNFSMount:
         request.addfinalizer(finalizer)
 
     @pytest.fixture()
-    def mount_noobaa_db_pod(self, request):
-        nfs_server = config.ENV_DATA.get("nb_nfs_server")
-        nfs_mount = config.ENV_DATA.get("nb_nfs_mount")
-        if not nfs_server or not nfs_mount:
-            pytest.skip(
-                f"NFS config not set: nb_nfs_server={nfs_server}, nb_nfs_mount={nfs_mount}"
-            )
+    def mount_noobaa_db_pod(self, request, nfs_config):
+        nfs_server, nfs_mount = nfs_config
 
         # scale down noobaa db stateful set
         helpers.modify_statefulset_replica_count(

--- a/tests/functional/object/mcg/test_db_nfs_mount.py
+++ b/tests/functional/object/mcg/test_db_nfs_mount.py
@@ -29,6 +29,13 @@ logger = logging.getLogger(__name__)
 class TestNoobaaDbNFSMount:
     @pytest.fixture()
     def mount_ngix_pod(self, request):
+        nfs_server = config.ENV_DATA.get("nb_nfs_server")
+        nfs_mount = config.ENV_DATA.get("nb_nfs_mount")
+        if not nfs_server or not nfs_mount:
+            pytest.skip(
+                f"NFS config not set: nb_nfs_server={nfs_server}, nb_nfs_mount={nfs_mount}"
+            )
+
         # try to mount the reesi004 nfs mount to nginx pod
         nginx_pod_data = templating.load_yaml(constants.NGINX_POD_YAML)
         nginx_pod_data["metadata"]["namespace"] = config.ENV_DATA["cluster_namespace"]
@@ -38,18 +45,15 @@ class TestNoobaaDbNFSMount:
         ] = "/var/nfs"
         nginx_pod_data["spec"]["volumes"][0]["name"] = "nfs-vol"
         nginx_pod_data["spec"]["volumes"][0]["nfs"] = dict()
-        nginx_pod_data["spec"]["volumes"][0]["nfs"]["server"] = config.ENV_DATA.get(
-            "nb_nfs_server"
-        )
-        nginx_pod_data["spec"]["volumes"][0]["nfs"]["path"] = config.ENV_DATA.get(
-            "nb_nfs_mount"
-        )
+        nginx_pod_data["spec"]["volumes"][0]["nfs"]["server"] = nfs_server
+        nginx_pod_data["spec"]["volumes"][0]["nfs"]["path"] = nfs_mount
         nginx_pod_data["spec"]["volumes"][0].pop("persistentVolumeClaim")
         nginx_pod = helpers.create_resource(**nginx_pod_data)
 
         helpers.wait_for_resource_state(
             nginx_pod,
             constants.STATUS_RUNNING,
+            timeout=300,
         )
         logger.info(f"Pod {nginx_pod.name} is created and running")
 
@@ -61,6 +65,13 @@ class TestNoobaaDbNFSMount:
 
     @pytest.fixture()
     def mount_noobaa_db_pod(self, request):
+        nfs_server = config.ENV_DATA.get("nb_nfs_server")
+        nfs_mount = config.ENV_DATA.get("nb_nfs_mount")
+        if not nfs_server or not nfs_mount:
+            pytest.skip(
+                f"NFS config not set: nb_nfs_server={nfs_server}, nb_nfs_mount={nfs_mount}"
+            )
+
         # scale down noobaa db stateful set
         helpers.modify_statefulset_replica_count(
             constants.NOOBAA_DB_STATEFULSET, replica_count=0
@@ -76,8 +87,8 @@ class TestNoobaaDbNFSMount:
         params = (
             '{"spec": {"template": {"spec": {"containers": [{"name": "db", "volumeMounts": '
             '[{"mountPath": "/var/nfs", "name": "nfs-vol"}]}], "volumes":[{"name": "nfs-vol", '
-            f'"nfs": {{"server": "{config.ENV_DATA.get("nb_nfs_server")}", '
-            f'"path": "{config.ENV_DATA.get("nb_nfs_mount")}"}}}}]}}}}}}}}'
+            f'"nfs": {{"server": "{nfs_server}", '
+            f'"path": "{nfs_mount}"}}}}]}}}}}}}}'
         )
         ocp_obj.patch(params=params, format_type="strategic")
         logger.info("Patched noobaa sts to mount nfs")

